### ISSUE TITLE
Block Styles button labels are truncated

### ISSUE THREAD

**Issue by erickyle** [WordPress/gutenberg#40331]
### Description

The block style button label is constrained to a small number of characters. Hovering over the block style button does reveal a preview with the full name, but this forces the user to hover over each of the block style buttons to reveal the full name. The built-in blocks and block styles tend to use single-word, short labels, so the issue may not be readily apparent with a stock configuration.

![image](https://user-images.githubusercontent.com/14015502/163280364-76e0e1b6-acad-47ff-af69-c01c9a34f19d.png)

It may be better to present the block style buttons as a single column with a much larger number of characters supported, versus the two-column button layout that is currently in use. Note that the screen capture below uses a a transparent variant of the `--wp-admin-theme-color--rgb` CSS custom property.

![image](https://user-images.githubusercontent.com/14015502/163282431-facd4f5e-42de-4f1e-b53b-792dbd8f678c.png)

This issue is related to #37039

### Step-by-step reproduction instructions

1. Insert a block that has block styles with long names associated with it.
2. Inspect the block styles in the settings panel. 

### Environment info

- WordPress 6.0-beta1-53168
- Twenty Twenty-Two theme active
- WordPress Beta Tester is the only plugin installed
- Gutenberg plugin is not installed

### Please confirm that you have searched existing issues in the repo.

Yes

### Please confirm that you have tested with all plugins deactivated except Gutenberg.

Yes

**Comment by jasmussen** [comment]
Good ticket, thank you, I'll be curious to hear what others think as well. I personally like that the button size encourages short and concise labels, and that the overall footprint of the style interface is balanced against design tools below, making the interface a set of complementary shortcuts rather than the primary block interface.

I wonder if there's a way to automatically adjust the interface size, such as if more than `n` styles are detected, it becomes a full-width dropdown? 🤔

**Comment by erickyle** [comment]
Glad to help. I would be fine with leaving it as-is if there was a way to override the truncation easily. I wonder instead of truncating the string before output, if it would not be better to truncate it after output using CSS.
https://css-tricks.com/snippets/css/truncate-string-with-ellipsis/

**Comment by mrwweb** [comment]
> I'll be curious to hear what others think as well. I personally like that the button size encourages short and concise labels

I do not like it and find that it's hard to write descriptive enough labels in the available space. I wouldn't say "encourage" but rather "enforce." I honestly think it should just be reverted to show the whole label, but some kind of override would be acceptable. If there is a huge concern about the space it takes up, using a select menu or a scrollable container would be an improvement over the current implementation (though a scrollable container would have its own downsides).

Since this is a developer opt-in feature, it feels like it should be left to the developer to make this decision. I agree that shorter labels are better, but the available space _really short_.

I spend a lot of time and through trying to write short labels and I can't even come close to fitting in the available space at least half the time. Here are some recent examples of styles that I don't think I could shorten any further that are truncated:

- Circle Gradient: Brick to Blue ("Circle Gradi…")
- Circle Gradient: Yellow to Teal  ("Circle Gradi…")
- Not Stretched ("Not Stretche..." - In this case the ellipsis truncates the word when it wouldn't have to without it!)
- Centered Not Stretched ("Centered Not…")
- Small Grayscale Logos ("Small Graysc…")
- Snug to Heading ("Snug to Head…")
- space-around Alignment ("space-around..." - See #42536)

Especially since it's not possible to combine Block Styles (#14598), it's more likely to need longer, composite names like the "Small Grayscale Logos" or Circle Gradient variation styles above.

Further, it's possible that the tooltip avoids a formal accessibility failure, but I think most of the logic in a few recent articles still applies.

From ["The Ballad of Text Overflow"](https://www.tpgi.com/the-ballad-of-text-overflow/):

> If text has been truncated with text-overflow, then this is a loss of content, and therefore an instant failure of 1.4.10.

See also: ["Text Overflow: Ellipsis Considered Harmful"](Text-overflow: ellipsis considered harmful).

**Comment by ciampo** [comment]
Agreed with Lena — select feels like the natural choice here, especially since we don't have any "radio cards" components yet.

**Comment by t-hamano** [comment]
> I think the wp-ui version would allow for the preview cards as well.

I did a little testing. Is this how it looks? @jasmussen What do you think?

https://github.com/user-attachments/assets/05103094-35c5-4fd7-ba06-379dc274a813

**Comment by jasmussen** [comment]
Thanks for exploring. It's a good exploration, and great to help us decide. For labels that are _short_ like in the example, this is arguably a worse experience however, than just the immediately available buttons. 

I guess it's difficult to discern whether the labels are long? The hope was that only when the labels are long, do they turn into a dropdown. 

This is probably a stretch, so don't take me too seriously, but would it be possible to turn it into a dropdown if you have more than four buttons, is that simpler?

To be clear: I'm not strongly objecting to any one specific path forward. I'm still in the phase of ensuring all corners are explored before we settle on a precise path forward. Ideas are welcome.

**Comment by t-hamano** [comment]
Previous issue: https://github.com/WordPress/gutenberg/issues/33933

The current UI has the advantage of allowing users to see at a glance how many block styles a block has and to change styles with a single click. However, considering consistency with the design system, an approach using ui/Select would be more sensible.

> but would it be possible to turn it into a dropdown if you have more than four buttons, is that simpler?

If we adopt this approach, I think the buttons should always occupy a single column. Allowing them to switch between one or two columns based on text length might make the UI unnecessarily complex.

**Comment by jasmussen** [comment]
The problem with a single column (which arguably would afford longer descriptions on its own) is that it takes up more vertical space, and thus pushes down valuable panels like colour, typography, border, etc. In that sense, if the only solution is always a dropdown, then that's what I'd support of the available options. But what if you only have two styles? Or what if you only have one? The thing about the select is that it is designed for a scale, both in allowing verbose titles but also amount of options, that is not always present, and when it isn't present, it can be a bit much to add that extra click. 

No strong opinions, but still feeling like that good unexplored and simple solution is still out there in the fog, waiting to be found.

**Comment by ciampo** [comment]
Maybe the best solution is to alternate between a `ToggleGroupControl` and `SelectControl` depending on the number of options and/or actual width required to render them?

But if we want to keep it simple, I agree that Select feels like the right component for the job

**Comment by cbirdsong** [comment]
I think a dropdown only makes sense if there are a lot of options. It feel pretty be silly to have one for the stock "Fill" and "Outline" options on the button block, for instance.

I also think the quick and dirty truncation removal is pretty good, especially if you could keep it to only two lines instead of one?

**Comment by jasmussen** [comment]
> I also think the quick and dirty truncation removal is pretty good, especially if you could keep it to only two lines instead of one?

Buttons-system-wise, is there a clean and nice way to do this?

**Comment by ciampo** [comment]
To make sure I'm following here, the latest proposal is to:

- Keep two columns when labels fit comfortably.
- Let an individual option occupy the full row when it needs more width.
- Clamp button labels on one line (or two at most). Truncate when they overflow

> > I also think the quick and dirty truncation removal is pretty good, especially if you could keep it to only two lines instead of one?
> 
> Buttons-system-wise, is there a clean and nice way to do this?

The `@wordpress/ui` Button allows its content to break to multiple lines, but:

- its height obviously increases, which could pose a challenge when some buttons become taller than others
- it doesn't handle truncation + tooltip out of the box (that is a conscious decision; relevant conversations: https://github.com/WordPress/gutenberg/issues/77391, https://github.com/WordPress/gutenberg/pull/78256 cc @mirka )

<img width="476" height="164" alt="Image" src="https://github.com/user-attachments/assets/dc8932d8-a45f-4a05-9962-c89c933ef15a" />

---

Here are my thoughts:
- We should reduce truncation as much as possible. Allowing labels to wrap to two lines and buttons to occupy a full row would both be improvements.
- A dynamic layout isn't completely out of the question, but we need to define its behavior precisely before deciding whether it is simple to implement. A few questions that come to mind:
  - How do we determine that a label needs a full row? It should depend on rendered width and it will need to respond to various external factors (eg translations, fonts, zoom, sidebar width...). If we end up needing to measure with JS, it may be complex and runtime expensive.
  - If a short label is followed by a long one, does the long button move to a full-width row? If so, does the short button remain alone at half width? Or expand to full width to fill the empty gap?
  - What happens to an odd final button? Does it remain half-width or expand? 
  - Should buttons in the same row always have equal heights? (relevant if we allow button labels to wrap on 2 lines)
- CSS intrinsic sizing may be able to handle some of this without JavaScript measurement. However, CSS still needs deterministic layout rules; it cannot answer the design questions above for us.
- The semantics on this widget should also be reconsidered, potentially making it a radio group (visually styled as buttons)

My concern is that we risk overcomplicating this feature.

**Comment by jasmussen** [comment]
Thanks for your deep response, much appreciated.

> To make sure I'm following here

To clarify specifically what I meant, I meant keep _everything_ as it exists today, but simply allow the buttons to wrap the text inside, like this but ideally nicer:

<img width="284" height="153" alt="Image" src="https://github.com/user-attachments/assets/11960e4c-e6f1-4c92-af6f-1eaec610fecf" />

It may not work, but if it does, it keeps one-click style changes, affords more verbose labels, and only pushes content in the sidebar down in the cases where you have a LOT of styles, in which case that might be desired after all. 

The main reason this could _not work_ is if the grid of buttons ends up entirely lopsided, with some buttons taller than others. That's probably the main challenge.

**Comment by ciampo** [comment]
Right! Yes, some buttons will be taller than others. We can mitigate by making sure that buttons in the same row have the same height? Would that be enough?

**Comment by jasmussen** [comment]
If it's relatively trivial to try, it might be, yes. 

The thing is: it would possibly address this issue, but it would also leave it up to theme-developers to editorially balance their style titles. I.e. with no risk of truncation, they could carefully tailor their titles to have all 1 or 2 or 3 lines of text, or be fine with differeing sizes.

**Comment by ciampo** [comment]
Allowing up to 2 lines of text and then truncating would be the better compromise?

**Comment by jasmussen** [comment]
None of my opinions are strong, and I defer to any others on this thread chiming in.

But I wouldn't truncate anything at all, and leave it completely to the theme-developer to editorially decide. If a followup issue materialises that shows a 10-line style button, I guess that's when we know the dropdown is the answer. Is that fair?

**Comment by ciampo** [comment]
Sounds good to me for a couple of reasons:

- we potentially avoid truncation (and tooltip) altogether, which is great
- we start simple, and can always revisit later if necessary.


The risks to monitor are:

- extemely long labels that look weird
- consumers implementing their own truncation in a non-accessible way (ie. without correct accessible label and tooltips)

**Comment by t-hamano** [comment]
> Allowing up to 2 lines of text and then truncating would be the better compromise?

I tested this. The code changes are very simple.

https://github.com/WordPress/gutenberg/compare/update/block-styles-ui-button?expand=1

<details><summary>Code to add block style variations for the Paragraph block from the browser console</summary>

```javascript
( () => {
	const BLOCK = 'core/paragraph';

	// The order is fixed, not shuffled, so that each grid row pairs labels of a
	// known length. `getRenderedStyles` prepends a "Default" style because none
	// of these is the default, so "Default" fills the first cell of row 1 and
	// every pair below lands on its own row.
	const styles = [
		// Row 1 (with "Default"): both fit on one line — the shortest row.
		{ name: 'test-1char', label: 'A' },

		// Row 2: a one-line label next to a two-line one, so the short button
		// is stretched to the height of its row.
		{ name: 'test-short', label: 'Bold' },
		{ name: 'test-two-lines', label: 'Large drop cap with serif type' },

		// Row 3: back to one line, so the height drops again between rows.
		{ name: 'test-short-2', label: 'Plain' },
		{ name: 'test-ja-short', label: '本文' },

		// Row 4: both run past two lines and end in an ellipsis.
		{
			name: 'test-long',
			label: 'Subtle background with rounded corners',
		},
		{
			name: 'test-very-long',
			label: 'Extra emphasised introduction paragraph for long form articles',
		},

		// Row 5: unbreakable words — they must wrap mid-word rather than widen
		// the column, which is what `overflow-wrap: anywhere` and
		// `minmax(0, 1fr)` are there for.
		{
			name: 'test-unbreakable',
			label: 'Donaudampfschifffahrtsgesellschaftskapitaenswitwe',
		},
		{
			name: 'test-url',
			label: 'https://example.com/very/long/path/to/a/block/style',
		},

		// Row 6: CJK line breaking and truncation.
		{
			name: 'test-ja-long',
			label: '長いラベルを持つ日本語のブロックスタイルバリエーション見本',
		},
		{ name: 'test-ja-short-2', label: '太字' },
	];

	wp.blocks.registerBlockStyle( BLOCK, styles );

	window.unregisterTestBlockStyles = () => {
		styles.forEach( ( { name } ) =>
			wp.blocks.unregisterBlockStyle( BLOCK, name )
		);
		delete window.unregisterTestBlockStyles;
	};

	console.log(
		`Registered ${ styles.length } styles on ${ BLOCK }. Run unregisterTestBlockStyles() to remove them.`
	);
} )();
```

</details> 

| Before | After |
|--------|--------|
| <img width="280" height="224" alt="Image" src="https://github.com/user-attachments/assets/f1b988be-b908-4411-9277-a4effce4f616" /> | <img width="280" height="219" alt="Image" src="https://github.com/user-attachments/assets/bbd56750-558a-4733-8f86-737dad49841a" /> |
| <img width="277" height="555" alt="Image" src="https://github.com/user-attachments/assets/2a94f0d6-b980-47a8-8354-9cdef69c2586" /> | <img width="278" height="595" alt="Image" src="https://github.com/user-attachments/assets/e753ad12-c38a-486f-aa86-001951133156" /> |

The key point is that the button height changes depending on the row, but personally, I find this acceptable.

**Comment by jasmussen** [comment]
Does it start to look very poor with 3 lines? If it isn't terrible, and depending on any others chiming in, it could be key.

**Comment by t-hamano** [comment]
> Does it start to look very poor with 3 lines?

I tried this. https://github.com/WordPress/gutenberg/pull/81833 The PR also includes accessibility improvements.

**Comment by jasmussen** [comment]
> I tried this. https://github.com/WordPress/gutenberg/pull/81833 The PR also includes accessibility improvements.

Thanks for trying that. Your examples are extreme (which is good!) with a simple "A" term on the left, and a three line row on the right. It does look a bit lopsided there. 

However, it's still the simplest possible solution, and it notably bends to the will of the theme developer and what they desire to do editorially: able to choose all in one line each, or two, or three lines. Mixing is possible too.

Because it is the simplest solution that fixes this issue, and absent anyone from this issue chiming in, I would be in favour of trying this as the first step.
